kvdb/src/lib.rs

```diff
@@ -167,9 +181,432 @@ pub trait KeyValueDB: Sync + Send {
 	fn restore(&self, new_db: &str) -> io::Result<()>;
 }
 
-/// Generic key-value database handler. This trait contains one function `open`. When called, it opens database with a
-/// predefined config.
-pub trait KeyValueDBHandler: Send + Sync {
-	/// Open the predefined key-value database.
-	fn open(&self, path: &Path) -> io::Result<Arc<KeyValueDB>>;
+
+/// Converts an error to an `io::Error`.
+pub fn other_io_err<E>(e: E) -> io::Error where E: Into<Box<std::error::Error + Send + Sync>> {
+	io::Error::new(io::ErrorKind::Other, e)
+}
+
+/// An abstraction over a concrete database write transaction implementation.
+pub trait WriteTransaction {
+	/// Insert a key-value pair in the transaction. Any existing value will be overwritten upon write.
+	fn put(&mut self, col: usize, key: &[u8], value: &[u8]) -> io::Result<()>;
+	/// Delete value by key.
+	fn delete(&mut self, col: usize, key: &[u8]) -> io::Result<()>;
+	/// Commit the transaction.
+	fn commit(self: Box<Self>) -> io::Result<()>;
+}
+
+/// An abstraction over a concrete database read-only transaction implementation.
+pub trait ReadTransaction {
+	/// Get value by key.
+	fn get(self: Box<Self>, col: usize, key: &[u8]) -> io::Result<Option<DBValue>>;
+}
+
+pub trait TransactionHandler {
+	// TODO: how to avoid boxing?
```

Comment:
This would require [existential types](https://rust-lang.github.io/rfcs/2071-impl-trait-existential-types.html) right?

Reply:
Probably, also by fixing https://github.com/rust-lang/rust/issues/51386

Reply:
…so I guess that for the time being boxing is as good as it gets?

Reply:
I was hoping someone could with a better design, which would remove this requirement. Also I don't know impactful this boxing is.